python/pyspark/sql/dataframe.py

```diff
@@ -3244,10 +3244,14 @@ def drop(self, *cols: "ColumnOrName") -> "DataFrame":  # type: ignore[misc]
             else:
                 raise TypeError("col should be a string or a Column")
         else:
-            for col in cols:
-                if not isinstance(col, str):
-                    raise TypeError("each col in the param list should be a string")
-            jdf = self._jdf.drop(self._jseq(cols))
+            if all(isinstance(col, str) for col in cols):
+                jdf = self._jdf.drop(self._jseq(cols))
+            elif all(isinstance(col, Column) for col in cols):
+                jdf = self._jdf
+                for col in cols:
+                    jdf = jdf.drop(col._jc)  # type: ignore[union-attr]
```

Comment:
Can we avoid looping here? This is super expensive in Spark SQL optmizer. Ideally we should add the signature of `def drop(colNames: Column*` in Scala side first, and PySpark side directlly invokes it.

Reply:
> Can we avoid looping here? This is super expensive in Spark SQL optmizer. 

Excellent remark. I tried to find a proper way before going with this approach. Please bear with me as the explanation is long-ish.

> Ideally we should add the signature of `def drop(colNames: Column*` in Scala side first, and PySpark side directlly invokes it.

I started out with this exact assumption but it doesn't hold ground due to runtime type erasure. I need some inputs from committers before I can make a sane choice.
1. Addition of `def drop(cols: Column*)` wouldn't work. `def drop(cols: String*)` already exists leading to double definition because they have same type after erasure: (cols: Seq)org.apache.spark.sql.Dataset
2. We could add `def drop(col: Column, cols: Column*)`. That helps on JVM already but on pyspark it doesn't work as `def drop(colNames: String*)` gets precedence. `self._jdf.drop(self._jseq(cols, _to_java_column))` will always pick up `def drop(cols: String*)` at runtime and lead to an error. This is actually worse than where we started.
3. Not recommended: We could change `def drop(colNames: String*)` to `def drop(colName:String, colNames: String*)` but then that is a breaking change for JVM. However, other bindings are pretty happy.
4. Compromise: We could add `def dropMultiple(cols: Column*)` and `def dropMultiple(col: String, cols: String*)`. This is a middle ground but at the cost of an addition to public API. I am not a big fan of adding public API as a compromise.

What would you suggest @HyukjinKwon ?

Reply:
>  pyspark it doesn't work as `def drop(colNames: String*)` gets precedence. `self._jdf.drop(self._jseq(cols, _to_java_column))` will always pick up `def drop(cols: String*)` at runtime and lead to an error.

Can you call `def drop(col: Column, cols: Column*).` is the input is column instance?

Reply:
I tried, py4j always picks up `drop(cols: String*)`.

Reply:
We have `DataFrame.select` that both takes column and strings. Py4J will look up all refelected methods when they try to call JVM methods so it's weird that it's unable to call that method.

Reply:
That is expected. Select has following overloads: `select(cols: Column*)` and `select(col: String, cols: String*)` as opposed to `drop(col:Column, cols: Column*)` and `drop(cols: String*)`.

Reply:
You can call `def drop(col1: Column, col2: Column*)` as below. Tested:

```python
df._jdf.drop(df.id._jc, sc._jvm.PythonUtils.toSeq([df.id._jc]))
```

Reply:
Clever to split the list into first column and tail, let me make some amends. Thanks!

Reply:
Welp, with that change it breaks existing `drop(col: Column)` public API for bindings. 
```
py4j.Py4JException: Method drop([class org.apache.spark.sql.Column]) does not exist
```
I have confirmed that to be the case for both python and sparkR.

Reply:
Applied the suggestions and also made changes to both R and python APIs. Waiting for the pipelines to finish.

Reply:
@HyukjinKwon with a very minor tweak the implementation looks sane again! We can go for round two of the reviews! Thanks for your help.